### ISSUE TITLE
Performance difference 

### ISSUE THREAD

**Issue by aditya-droid** [AlexeyAB/darknet#4336]
Hi, 
First of all I would like to thank you for your amazing work on this repository especially for insights on working with YOLO. 

I have a question and I have tried to look for it on the internet but I am not convinced with the answers that I found. I am building a production level model and I am yet to acknowledge your work on the modification of darknet and YOLO. I would like to know the performance difference between this repository's implementation and that of the original pjreddie's implementation. 

Thanks for you help. 

- Aditya

**Comment by AlexeyAB** [comment]
It depends on model, task and hardware.

Training 2-3x faster
Detection on video 1.2x - 2x faster
Forward inference 1.2x faster
Accuracy +~20-40% on new models
Are supported models: EfficientNet, MixNet, TridentNet, Gaussian_yolo + GIoU, ...: https://github.com/AlexeyAB/darknet/projects/1

**Comment by aditya-droid** [comment]
I am implementing an intelligent traffic management system using YOLOv3. I am getting okay results not that great due to the low resolution of videos, but still even if high resolution they are not state of the art. I have decided to go with Yolov3 due to its fps on video processing. 

The results should be better on your version as you have mentioned. 

So you mean by supported models is that they are combined with Yolo or are they implemented on darknet framework without yolo? 

I do have 2 more questions:
1) Which model gives out the best mAP? 
2) Does the above models impact the processing fps speed on videos?

**Comment by AlexeyAB** [comment]
> Which model gives out the best mAP?

It depends on relative sizes of objects.

> Does the above models impact the processing fps speed on videos?

Yes. Some of models faster and more accurate, some of models more accurate but slower.

* What mAp do you get on Training and Validation datasets?
* What network size (width= height= )do you use in cfg-file?

* Show console output and screenshot of cloud of point:
`./darknet detector calc_anchors data/obj.data -num_of_clusters 9 -width 416 -height 416 -show`

* attach your cfg-file

**Comment by aditya-droid** [comment]
I haven't checked the mAP results for that. My network size is 416x416. I have the same cfg-file (yolov3.cfg)

So in your opinion which is a better model for a video demonstration of traffic? TridentNet or PAN or LSTM or Gaussian + GloU?

**Comment by AlexeyAB** [comment]
What is your video resolution?

Try this 
[yolo_v3_tiny_pan5_matrix_gaussian_GIoU_aa_ae_mixup.cfg.txt](https://github.com/AlexeyAB/darknet/files/3869623/yolo_v3_tiny_pan5_matrix_gaussian_GIoU_aa_ae_mixup.cfg.txt)

**Comment by aditya-droid** [comment]
My video resolution is 1080x720

**Comment by aditya-droid** [comment]
What about the weights for the same? I do test this out normally right? The way I do all of this with yolov3.cfg?

**Comment by AlexeyAB** [comment]
Yes, all the same, except:
* train without pre-trained weights
* set burn_in=1000
* set `filters  = (classes + 8 + 1) * <numbers in mask>`


All the same, set max_batches=, width=, height=, classes= ... as in your previous cfg
And use the latest version of Darknet from this repository.

**Comment by aditya-droid** [comment]
Alright, I will train the network using this. How do I get mAP for the video that I will be processing? I will training on images of cars, buses, trucks, motorcycles, cycles, auto-rickshaws, etc. Then I run the trained model on a traffic video and check out the detections. 

Do I need to calculate anchors for this?

**Comment by AlexeyAB** [comment]
Use default anchors.

Check the mAP on Training and Validation datasets.

If you want to check the mAP on video, then use  https://github.com/AlexeyAB/Yolo_mark to get frames from videofile (save each N frame, in example N=10), you can use this command:

on Windows: yolo_mark.exe data/img cap_video test.mp4 10
on Linux: ./yolo_mark x64/Release/data/img cap_video test.mp4 10

then label them by using Yolo_mark: 
on Windows: yolo_mark.exe data/img data/valid.txt data/obj.names
on Linux: ./yolo_mark x64/Release/data/img x64/Release/data/valid.txt x64/Release/data/obj.names

And specify `valid=valid.txt` in your `obj.data` file then run `./darknet detector map obj.data ...` https://github.com/AlexeyAB/darknet#when-should-i-stop-training

**Comment by minhtcai** [comment]
> Yes, all the same, except:
> 
>     * train without pre-trained weights
> 
>     * set burn_in=1000
> 
>     * set `filters  = (classes + 8 + 1) * <numbers in mask>`
> 
> 
> All the same, set max_batches=, width=, height=, classes= ... as in your previous cfg
> And use the latest version of Darknet from this repository.
> 
> And train **at least 7000 iterations**.

I thought you said 
For [Gaussian_yolo] should be:
filters = (classes + 8 + 1) * <numbers in mask>

instead of (for [yolo]):
filters = (classes + 4 + 1) * <numbers in mask>

**Comment by aditya-droid** [comment]
I am having trouble compiling darknet on Ubuntu with the flag opencv set to 1. 

`Perhaps you should add the directory containing `opencv.pc'
to the PKG_CONFIG_PATH environment variable
No package 'opencv' found
./src/image_opencv.cpp:16:10: fatal error: opencv2/imgproc/imgproc.hpp: No such file or directory
 #include <opencv2/imgproc/imgproc.hpp>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~`

I do have opencv4 installed: 
`$pkg-config --modversion opencv4
$4.1.2`

I am unable to find a solution for it. Could you help me out with this?

**Comment by AlexeyAB** [comment]
@aditya-droid How did you install OpenCV? Did you use Cmake?

**Comment by aditya-droid** [comment]
Yes, I installed OpenCV and used make to build the opencv and then sudo make install.

**Comment by AlexeyAB** [comment]
Try to build Darknet by using Cmake
`cmake . & make`

**Comment by aditya-droid** [comment]
Yes, was able to compile. Thanks. Will be testing out the configuration files you sent.

**Comment by nyj-ocean** [comment]
@AlexeyAB 
I found that `yolo_v3_tiny_pan5_matrix_gaussian_GIoU_aa_ae_mixup.cfg` do not use pre-trained weights(`yolov3-tiny.conv.15`) to train as shown in https://github.com/AlexeyAB/darknet/issues/4336#issuecomment-556020918 like following

> Yes, all the same, except:
>     train without pre-trained weights
>     set burn_in=1000

- Is it due to the` feature extractor(like darknet53 in yolov3 or darknet19 in yolov2)` in `yolo_v3_tiny_pan5_matrix_gaussian_GIoU_aa_ae_mixup.cfg` is different from `yolov3-tiny.cfg ` like following ?

> #### deformation-block
> 
> [maxpool]
> size=2
> stride=2
> 
> [convolutional]
> batch_normalize=1
> filters=8
> size=5
> stride=1
> pad=1
> activation=swish
> 
> [route]
> layers=-3
> 
> [maxpool]
> size=2
> stride=1
> stride_x=2
> stride_y=1

- I found that the `feature extractor` of  `yolo_v3_tiny_pan3.cfg` is similar to `yolov3-tiny.cfg`

so could I use  `pre-trained weights(yolov3-tiny.conv.15)`  to train `yolo_v3_tiny_pan3.cfg`

**Comment by AlexeyAB** [comment]
@nyj-ocean 

> Is it due to the feature extractor(like darknet53 in yolov3 or darknet19 in yolov2) in yolo_v3_tiny_pan5_matrix_gaussian_GIoU_aa_ae_mixup.cfg is different from yolov3-tiny.cfg like following ?

Yes

> so could I use pre-trained weights(yolov3-tiny.conv.15) to train yolo_v3_tiny_pan3.cfg

Yes

**Comment by aditya-droid** [comment]
I was trying to train with the above mentioned .cfg file but it gives me an assertion error. The error is visible in the screenshot. Also, what do you mean by masks? How do I know what is the number of masks? 

![image](https://user-images.githubusercontent.com/55411733/73262045-5bba0f00-41f3-11ea-9ed2-73aff641f915.png)
